mono/metadata/cominterop.c

```diff
@@ -3699,32 +3669,46 @@ mono_cominterop_get_com_interface (MonoObject *object, MonoClass *ic, MonoError
 			return NULL;
 		klass = mono_object_class (object);
 		if (!mono_class_is_transparent_proxy (klass)) {
+			if (icall)
+				g_assert_not_reached ();
 			mono_error_set_invalid_operation (error, "Class is not transparent");
 			return NULL;
 		}
 
 		real_proxy = ((MonoTransparentProxy*)object)->rp;
 		if (!real_proxy) {
+			if (icall)
+				g_assert_not_reached ();
 			mono_error_set_invalid_operation (error, "RealProxy is null");
 			return NULL;
 		}
 
 		klass = mono_object_class (real_proxy);
 		if (klass != mono_class_get_interop_proxy_class ()) {
+			if (icall)
+				g_assert_not_reached ();
 			mono_error_set_invalid_operation (error, "Object is not a proxy");
 			return NULL;
 		}
 
 		if (!((MonoComInteropProxy*)real_proxy)->com_object) {
+			if (icall)
+				g_assert_not_reached ();
```

Comment:
Same. Assert here, move check to caller.

Reply:
I don’t understand this suggestion. I formed it, here and elsewhere, to preserve exact same meaning & maximum reuse and I think it is pretty clear. Nothing is reordered for example. It is possible some subtle semantic changes are ok, like errors instead of not_reached, but I erred totally toward preserving semantics. Error_init could also be shared.

Reply:
Hm. well fair enough.

What about

```c
g_assert_message (!icall, "real_proxy->com_object can't be NULL in COM icall");
```

or something similar to replace `if (!icall) g_assert_not_reached()` with a slightly more informative crash message.

Reply:
Agreed. g_assert_not_reached should be banned, along with g_assert(FALSE). Let me give it a try.

Reply:
g_assert_not_reached has other problems in that Visual C++ will generate bad code sometimes when it is hit, like a switch default (__assume(0) that is). Maybe gcc/clang similar. assert is at least a well defined stop that the compiler does not overoptimize without its own proof, instead of trusting the programmer.